src/sync/impl/apple/network_reachability_observer.cpp

```diff
@@ -0,0 +1,127 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync/impl/apple/network_reachability_observer.hpp"
+
+using namespace realm;
+
+namespace {
+
+static NetworkReachabilityStatus reachability_status_for_flags(SCNetworkReachabilityFlags flags)
+{
+    if (!(flags & kSCNetworkReachabilityFlagsReachable)) {
+        return NotReachable;
+    }
+
+    NetworkReachabilityStatus status = NotReachable;
+
+    if (!(flags & kSCNetworkReachabilityFlagsConnectionRequired)) {
+        status = ReachableViaWiFi;
+    }
+
+    if ((flags & kSCNetworkReachabilityFlagsConnectionOnTraffic) &&
+        !(flags & kSCNetworkReachabilityFlagsInterventionRequired)) {
+        status = ReachableViaWiFi;
+    }
+
+#if TARGET_OS_IPHONE
+    if (flags & kSCNetworkReachabilityFlagsIsWWAN) {
+        status = ReachableViaWWAN;
+    }
+#endif
+
+    return status;
+}
+
+static void reachability_callback(SCNetworkReachabilityRef, SCNetworkReachabilityFlags, void* info)
+{
+    auto callback = reinterpret_cast<std::function<void ()> *>(info);
+    (*callback)();
+}
+
+} // (anonymous namespace)
+
+NetworkReachabilityObserver::NetworkReachabilityObserver(util::Optional<std::string> hostname,
+                                                         std::function<void (const NetworkReachabilityStatus)> handler)
+: m_callback_queue(dispatch_queue_create("io.realm.sync.reachability", nullptr))
```

Comment:
Can you make the second argument explicitly `DISPATCH_QUEUE_SERIAL`, just for better readability?